### ISSUE TITLE
Provisioning of the image fails. cmdline.txt not found

### ISSUE THREAD

**Issue by strausmann** [Tech-Byte-Tips/RPi-PXE#1]
Hello, 

I am currently testing the PXE server. I have created an image of a Pi4 with Ubuntu Server 22.04.4 LTS. No problems so far.
If I now want to use the image in front of my Pi, the data is copied, but unfortunately the customization of the cmdline.txt does not work at the end.

```bash
Changing the /boot/cmdline.txt to boot from the NFS directory ...
sed: can't read /NFSPXE/boot/1f5a4844/cmdline.txt: No such file or directory
```

When I check the structure on the Pi and the NFS for boot, the cmdline.txt is not directly in the boot folder but in the folder /boot/firmware

![image](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/4519228/013fa370-efaf-4d1a-8b4e-2c27d77c6e54)

![image](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/4519228/d3943bab-3905-44cb-925b-19dfb72ed3dc)

**Comment by PREngineer** [comment]
I need to rebuild my Raspberry Pi setup.  Please allow me some time until I can get everything up again.  I will come back and reply then.

**Comment by strausmann** [comment]
I looked around a lot yesterday. If you insert the SD card into a card reader, the files are directly in the root of the boot partition and there is no firmware folder. But if you booted from the SD card, there is this firmware folder. I'll start a Raspberry Pi OS Lite later and see if it behaves like that there.

**Comment by strausmann** [comment]
> The error that you reference means that when you made the image of the OS, the file was not in the location that it is expected to be in Raspbian. My script does not move files from where they were originally stored when you created the image.

I didn't expect that your script would shift anything. I took a closer look at it and looked at the part for the proselytizing. I also wanted to install an if chck if there is a firmware folder in the boot folder. but unfortunately it just didn't work. My check just didn't want to work.

**Comment by strausmann** [comment]
This is now a Raspberry PI OS Lite (64bit) on a Raspi 4. In the /boot/ there is also the firmware folder, which is not so directly on the SD card. The only difference is that the cmdline.txt and config.txt were linked with a symlink from the /boot/firmware folder to /boot.

![image](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/4519228/86bd1ff7-16f7-4fbb-b0b6-7cc3a9c7e9f8)

**Comment by strausmann** [comment]
But not as easy as expected. With the image of Raspberry Pi OS on the NFS server, the symlinks are missing in the backup of the /boot 

![image](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/4519228/30b89e95-78d4-44f4-a663-1c68e61913ef)

**Comment by MisterDeeds** [comment]
I would also be very interested in a solution. I think the problem arises with the new Raspbian version based on Debian 12. The copying process of the files works perfectly.

<img width="1042" alt="View" src="https://github.com/Tech-Byte-Tips/RPi-PXE/assets/8822811/b3b8e503-5adc-4f42-99da-33380e28ad38">

The structure does not appear to be known under PXE:

![IMG_4527](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/8822811/786c47d8-485d-40a1-9c3a-a19b2acad1ac)

**Comment by BurmazDaniel** [comment]
> I would also be very interested in a solution. I think the problem arises with the new Raspbian version based on Debian 12. The copying process of the files works perfectly.
> 
> <img alt="View" width="1042" src="https://private-user-images.githubusercontent.com/8822811/283061714-b3b8e503-5adc-4f42-99da-33380e28ad38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwNzA4ODcsIm5iZiI6MTcwMjA3MDU4NywicGF0aCI6Ii84ODIyODExLzI4MzA2MTcxNC1iM2I4ZTUwMy01YWRjLTRmNDItOTlkYS0zMzM4MGUyOGFkMzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDhUMjEyMzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTgwNDFkZGY0YTcwZDJkYzM4NDYyZGFkZjQ1MWNkYWYyNDRjYmNmOGY5NDQyNTQwNDVmYmJkZDJjMzk1YjAyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.bUPQX_VyZwhwcaZCSchNZrG2F1rORlkvJ1TynVvykPE">
> The structure does not appear to be known under PXE:
> 
> ![IMG_4527](https://private-user-images.githubusercontent.com/8822811/283062262-786c47d8-485d-40a1-9c3a-a19b2acad1ac.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIwNzA4ODcsIm5iZiI6MTcwMjA3MDU4NywicGF0aCI6Ii84ODIyODExLzI4MzA2MjI2Mi03ODZjNDdkOC00ODVkLTQwYTEtOWMzYS1hMTliMmFjYWQxYWMuSlBHP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDhUMjEyMzA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTVkMzhkYzViYzdhOTZiYzFiNTM2NjYyYTg5Y2YyOWQyYWZiY2I5YmRjOTBjMjI5NTU3NjdjYjA0OTk0YmU5ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gixSCRj_5pJgZNhh6UEYntYNxAIR6q4sghxWST8au-0)

I have the same problem with the new raspberry pi os (debian bookworm). I tried to place all files where rpi's bootloader wants to but unfortunately these files are nowhere to be found eeprom.sig, recover4.elf and recovery.elf. I think is has to do something with secure boot.

**Comment by PREngineer** [comment]
Hi, my apologies to all.  I've been super busy with work and I have not had the time to sit down and work on this.  I first need to rebuild my whole raspberry pi lab and then I'll get to work on this.  Don't worry, I will be updating it as I will definitely need it too once I rebuild my setup.  Please just pay attention to your e-mail once I make an update to this issue.

**Comment by strausmann** [comment]
For me, the topic was a few days ago again. But I think it's enough to put a few symlinks in the root of the image from the firmware folder.

**Comment by abeksis** [comment]
https://imgur.com/a/SrMKrsb it shows that more files are missing. I tried several times, and now I also tried another Ubuntu version.

**Comment by PREngineer** [comment]
Hi all!

I was trying to work on this today and found the same issue when trying to image/provision a pi using the latest Raspbian.
I can't find where to get these files, except the start.elf and start4.elf and I am not sure if they should all just be put in the base of each pi's boot directory.

I've placed a question in the raspberry pi forums to see if anybody has any clue as to where to get the files.

I'll keep you updated on the progress.

**Comment by PREngineer** [comment]
Copying everything from the firmware folder to the root (boot) for that serial number allowed it move farther in the boot process.
Now I'm getting a weird set of messages saying 

"Gave up waiting for root file system device."
"ALERT!  PARTUUID=5180406e-02 does not exist.  Dropping to a shell!"

At least is a little bit of progress.

**Comment by PREngineer** [comment]
Progress:

Made sure to copy the contents of the firmware folder to the root of the boot filesystem before attempting to provision a pi.

The script was successful in provisioning the pi and this time the cmdline.txt was pointing to the NFS share.

However, it seems that the new version, doesn't support NFS v4.1.  I am trying to figure out which version I can use to have it working.

**Comment by PREngineer** [comment]
https://forums.raspberrypi.com/viewtopic.php?t=357877

Seems like Bookworm has bugs that are causing this problem.

I would not recommend anybody to use Bookworm for PXE/NFS booting yet.

**Comment by brynjolm** [comment]
Im not sure if its solely based on bookworm. I think its the implementation of raspbian with regards to /boot. I did a lot of testing today and have managed to run the script albeit modified. It has something to do with the way the boot folder in raspberry pi is structured. On bullseye and older distros, raspberry pi has a bootloader setup of /boot and cmdline.txt or config.txt. Based on the newer updates from raspbian and from what i understand is the bootloader looks for the files inside /boot/firmware which is now the default location for the files. The important files are symlinked from /boot to /boot/firmware. I have managed to make this work by imaging the pi then physically copying from the /images export first the files inside /boot/firmware onto /boot/. Then re running the script to provision the pi. Otherwise the script will not be able to write the necessary cmdline.txt file.

**Comment by PREngineer** [comment]
That's precisely what I am trying to do too.

Copied everything from /boot/firmware to /boot.  Then switched the cmdline.txt to use:

console=serial0,115200 console=tty1 root=/dev/nfs nfsroot=${IP}:/volume1/PXE/filesystems/#{serial},vers=3 rw ip=dhcp rootwait elevator=deadline

I managed to get a Pi 4B booting successfully now.  But I see failures to load "systemd-remount-fs.service." Works when you plug the power into the pi to boot for the 'first time'.  However, if you run the 'sudo reboot now' command it won't boot after.

Seems very unstable right now to try to PXE/NFS boot.

**Comment by brynjolm** [comment]
I agree. I managed to get it stable enough to load everything but ssh-breaks and nfs is very spotty. I looked around and there are no distros that have the old bootloader setup. Everything has been updated and moved to /boot/firmware. I am toying with the idea that it could be possible to do a hybrid approach to this where the /boot folder remains in the sd card and everything gets moved to nfs. I saw an old post on the raspbian forums a few hours ago. But im too tired to test now.

**Comment by PREngineer** [comment]
https://www.raspberrypi.com/documentation/computers/config_txt.html

Reading up on the config file to see if it helps solve the issues.

**Comment by PREngineer** [comment]
The only thing that seems to work so far is using the /boot/firmware as the /boot partition from now on.

It does have some issues when loading the kernel modules, but in general, it boots and works.

![image](https://github.com/Tech-Byte-Tips/RPi-PXE/assets/6334069/7583e73e-1417-430c-814d-b774ef221712)

Until we have a better solution, this workaround will be applied to the main branch.

Currently testing the changes.

**Comment by PREngineer** [comment]
I was able to provision a Raspberry Pi 4B using Bookworm using these changes that I just pushed.
It's just a workaround for now.  There is one kernel module that fails to load, per the previous comment but otherwise, works fine.
Will be closing this issue for now.  I will keep monitoring the forums to see when a stable solution comes out.  Especially since I had to downgrade to NFS v3.
If anybody in this thread sees anything before me, I would appreciate it if you let me know through a new Issue mentioning the solution URL.
